opentracing/index.js

```diff
@@ -0,0 +1,165 @@
+const Agent = require('../lib/agent')
+const Transaction = require('../lib/instrumentation/transaction')
+
+class Tracer extends ot.Tracer {
+  constructor (agent) {
+    super()
+    if (agent && agent instanceof Agent) {
+      this._agent = agent
+    } else {
+      this._agent = require('../').start(agent)
+    }
+  }
+
+  _startSpan (name, fields) {
+    let context = null
+
+    if (fields.references && fields.references.length > 0) {
+      if (fields.references.length > 1) this._agent.logger.debug('Elastic APM OpenTracing: Unsupported number of references to _startSpan:', fields.references.length)
+
+      fields.references.some(ref => {
+        const type = ref.type()
+        if (!context && type === Tracer.REFERENCE_FOLLOWS_FROM) {
+          context = ref
+        } else if (type === Tracer.REFERENCE_CHILD_OF) {
+          context = ref
+          return true
+        }
+      })
+
+      context = context.referencedContext()._elasticContext
+    }
+
+    const span = this._agent.currentTransaction === null || this._agent.currentTransaction.ended // TODO: Is this right?
+      ? this._agent.startTransaction(name, null, context)
+      : this._agent.startSpan(name, null, context)
+
+    return span ? new Span(this, span) : Noop.span // TODO: What happens when a user tries to use the context of a Noop span? Maybe the Noop span context should be the context of the transaction?
+  }
+
+  // TODO: Align with other agents about TEXT/BINARY format
+  // TODO: Maybe wait implementing the binary format until: https://github.com/w3c/trace-context/tree/master/trace_context#binary-format
+  // - On the other hand, as we're not putting this is a special location, we might not have to wait on Trace Context - similar to how we don't wait for the HTTP definition when using a custom header prefix
+  _inject (spanContext, format, carrier) {
+    switch (format) {
+      case Tracer.FORMAT_TEXT_MAP:
+      case Tracer.FORMAT_HTTP_HEADERS:
+        carrier['elastic-apm-traceparent'] = spanContext.toString()
+        break
+      case Tracer.FORMAT_BINARY:
+        carrier.buffer = spanContext._elasticContext.buffer
+        break
+    }
+  }
+
+  _extract (format, carrier) {
+    let ctx
+
+    switch (format) {
+      case Tracer.FORMAT_TEXT_MAP:
+      case Tracer.FORMAT_HTTP_HEADERS:
+        ctx = ElasticTraceContext.fromString(carrier['elastic-apm-traceparent'])
+        break
+      case Tracer.FORMAT_BINARY:
+        ctx = new ElasticTraceContext(carrier.buffer)
+        break
+    }
+
+    return ctx ? new SpanContext(ctx) : null
+  }
+}
+
+class Span extends ot.Span {
+  constructor (tracer, span) {
+    super()
+    this.__tracer = tracer
+    this.__context = null
+    this._span = span
+    this._isTransaction = span instanceof Transaction
+  }
+
+  _context () {
+    return this.__context
+      ? this.__context
+      : (this.__context = new SpanContext(this._span.context))
```

Comment:
this was really hard to find while doing the test.. should be `this._span._context`

Reply:
Good catch!